atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java

```diff
@@ -71,10 +77,28 @@ public boolean requiresJointConsensus() {
     return !oldMembers.isEmpty();
   }
 
+  /**
+   * @return a set of all members in the configuration. During joint consensus where a member can be
+   *     in both the old and new configuration, the member with the higher {@link RaftMember.Type
+   *     type} is used.
+   */
   public Set<RaftMember> allMembers() {
-    final var all = new HashSet<RaftMember>(oldMembers.size() + newMembers.size());
-    all.addAll(newMembers);
-    all.addAll(oldMembers);
-    return all;
+    return new HashSet<>(
+        Stream.concat(newMembers.stream(), oldMembers.stream())
+            .collect(
+                Collectors.toMap(
+                    RaftMember::memberId,
+                    Function.identity(),
+                    BinaryOperator.maxBy(Comparator.comparing(RaftMember::getType))))
```

Comment:
:question: Is this comparing by ordinal?

Reply:
Yes, the type is an enum and inherits the default `Comparable` from enums. We already rely on the ordering of enum variants in several other places so I think that's okay.